lib/devicetrust/native/device_darwin.m

```diff
@@ -0,0 +1,249 @@
+  }
+  if (err) {
+    CFRelease(err);
+  }
+  return res;
+}
+
+int64_t DeviceKeyGet(PublicKey *pubKeyOut) {
+  CFDictionaryRef attrs = NULL;
+  CFDataRef appTagData = NULL; // managed by attrs
+  NSString *appTag = NULL;     // managed by ARC
+  SecKeyRef privKey = NULL;    // managed by attrs
+  SecKeyRef pubKey = NULL;
+  CFErrorRef err = NULL;
+  CFDataRef pubKeyRep = NULL;
+
+  int64_t res = findDeviceKey(@YES /* retAttrs */, (CFTypeRef *)&attrs);
+  if (res != errSecSuccess) {
+    goto end;
+  }
+
+  appTagData = CFDictionaryGetValue(attrs, kSecAttrApplicationTag);
+  if (!appTagData) {
+    res = kErrNoApplicationTag;
+    goto end;
+  }
+  appTag = [[NSString alloc] initWithData:(__bridge NSData *)appTagData
+                                 encoding:NSUTF8StringEncoding];
+
+  privKey = (SecKeyRef)CFDictionaryGetValue(attrs, kSecValueRef);
+  if (!privKey) {
+    res = kErrNoValueRef;
+    goto end;
+  }
+
+  pubKey = SecKeyCopyPublicKey(privKey);
+  if (!pubKey) {
+    res = kErrCopyPubKeyFailed;
+    goto end;
+  }
+
+  pubKeyRep = SecKeyCopyExternalRepresentation(pubKey, &err);
+  if (!pubKeyRep) {
+    res = CFErrorGetCode(err);
+    goto end;
+  }
+  copyPublicKey([appTag UTF8String], pubKeyRep, pubKeyOut);
+
+end:
+  if (pubKeyRep) {
+    CFRelease(pubKeyRep);
+  }
+  if (err) {
+    CFRelease(err);
+  }
+  if (pubKey) {
+    CFRelease(pubKey);
+  }
+  if (attrs) {
+    CFRelease(attrs);
+  }
+  return res;
+}
+
+int64_t DeviceKeySign(Digest digest, Signature *sigOut) {
+  SecKeyRef privKey = NULL;
+  NSData *data = NULL; // managed by ARC
+  CFErrorRef err = NULL;
+  CFDataRef sig = NULL;
+
+  int64_t res = findDeviceKey(@NO /* retAttrs */, (CFTypeRef *)&privKey);
+  if (res != errSecSuccess) {
+    goto end;
+  }
+
+  data = [NSData dataWithBytes:digest.data length:digest.data_len];
+  sig = SecKeyCreateSignature(privKey,
+                              kSecKeyAlgorithmECDSASignatureDigestX962SHA256,
+                              (CFDataRef)data, &err);
+  if (!sig) {
+    res = CFErrorGetCode(err);
+    goto end;
+  }
+
+  sigOut->data_len = CFDataGetLength(sig);
+  sigOut->data = calloc(sigOut->data_len, sizeof(uint8_t));
+  memcpy(sigOut->data, CFDataGetBytePtr(sig), sigOut->data_len);
+
+end:
+  if (sig) {
+    CFRelease(sig);
+  }
+  if (err) {
+    CFRelease(err);
+  }
+  if (privKey) {
+    CFRelease(privKey);
+  }
+  return res;
+}
+
+int64_t DeviceCollectData(DeviceData *out) {
+  CFStringRef cfSerialNumber = NULL; // managed via bridge
+  NSString *serialNumber = NULL;     // managed by ARC
+  int64_t res = 0;
+
+  io_service_t platformExpert = IOServiceGetMatchingService(
+      0 /* mainPort */, IOServiceMatching("IOPlatformExpertDevice"));
+  if (!platformExpert) {
+    res = -1;
+    goto end;
+  }
+
+  cfSerialNumber = IORegistryEntryCreateCFProperty(
+      platformExpert, CFSTR(kIOPlatformSerialNumberKey), kCFAllocatorDefault,
+      0 /* options */);
+  if (!cfSerialNumber) {
+    res = -1;
```

Comment:
Shouldn't we set a different error code for each error? If this function returns `-1` we won't know where the error comes from.

Reply:
Good call, done.